### DISCUSSION TITLE
What is the best way to implement beatmap export package without having to save the beatmap?

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#23490] [Ideas] [answered]
This was a little discussion on the [discord](https://discord.com/channels/188630481301012481/188630652340404224/1106268194043334748) between me and @bdach and it turned out inconclusive so i post it here.

Conversation:

@OliBomby:
>I want to make a legacy beatmap export option which would make use of the BezierConverter to make the sliders compatible with osu stable (long requested feature). Right now the export flow first saves the beatmap to realm and then exports the realm live model thingy to archive.
Since we don't actually want to overwrite our beatmap with the converted sliders version, i thought it would be good to somehow export without writing to realm. Is this possible? What would be the best flow for this problem?

@bdach:
>export shouldn't be using the save function at all ideally if the beatmap hasn't changed 
it is already a source of different issues (such as export actually overwriting the map)
the actual saving part is easy, just call the encoder. the bigger issue is how to make work in editor
possibly we may want to ask the user to save the beatmap before exporting explicitly if it is dirty, and only allow exports of clean/saved beatmaps
is what my opinion would be i guess

@OliBomby:
>How I expect it to work is that if I use the export package feature in the editor, it would export the version which includes any unsaved changes without actually saving the beatmap

@bdach:
>i guess that also works
>in theory then it should be simple enough to throw away the current flow that (ab)uses save, and instead call the beatmap encoder directly on the in-memory beatmap 
it would be nice if you managed to split the relevant parts of save that tackle encoding off and reuse in both places
there's also the part where you'll likely need a temporary directory to copy beatmap files off to and zip them from there but that should be simple enough

@OliBomby:
>the exporting flow with realm goes quite deep. All the legacy exporters are made to only work with realm files. What i'm trying to do is essentially export the beatmap set realm object with a single file different

@bdach:
>that is the case now, but it may not be in the future
see: future support for editing storyboard
i think it would be beneficial to get a flow in for being able to extricate the files from the realm-bound structure and export the entire beatmap from another place rather than apply workarounds such as substituting one file 

@OliBomby:
>What is this 'another place' then exactly? Should the exporter be able to export some collection of files which are not just files in the Realm but also files in arbitrary places on the disk and files in memory?

@bdach:
>well no you'd do it as part of the export operation
sketch in my head is
> - iterate over all beatmap files already mapped in realm
> - copy them out from lazer file store to temp directory on disk
> - encode currently edited beatmap in its current state to that same directory on disk
> - zip directory, put zip in the exports folder, present zip to user
>
> steps 1 & 2 will be especially useful for the future - both for storyboard editor once that happens, but also for other flows such as https://github.com/ppy/osu/issues/12029

@OliBomby:
>is that copying of lazer file store to temp directory disk actually good? I thought it was the whole purpose of Realm to handle storing files

@bdach:
>well i mean it's only for the sake of the export
i guess you can also create the archive in place just name the archive entries correctly

@OliBomby:
>sketch in my head rn is
> - create a new temporary copy of the BeatmapSetInfo
> - encode the currently edited beatmap to this copy
> - export the copy using the existing package exporters
> original beatmap set should remain unchanged right

@bdach:
>i mean i guess this works but i don't like it too much
it will be very finicky to make happen

@OliBomby:
>its essentially the same as your sketch but using realm instead of a directory

@bdach:
>well no, because you'd need to create a whole bunch of realm objects that you'd then need to clean up eagerly or mark as soft deleted
it's not just the BeatmapSetInfo that you'd need to copy. that has references to BeatmapInfo, BeatmapMetadata, BeatmapDifficulty
all are realm-persisted models
also encoding isn't something that happens into realm, it happens into the file store. the .osu files are also not in realm, they are in files/ under the .osu file's SHA
which is another thing you'd have to write out and then clean up

@OliBomby:
>are there no recursive deep copy and delete helpers which let you just create and clean up these realm-persisted models easily?

@bdach:
>i suppose you could .Detach() the beatmap set, reset its guid, and reinsert it to realm. it will probably work. but still, i don't overly like that solution
i would need to check if the detach detaches the entire object graph, or just the BeatmapSetInfo however
i don't remember from the top of my head. the automapper usage suggests that it might

@OliBomby:
>it seems like the the beatmap set in the save flow is already detached, this solution might not require that much change

And then we agreed to disagree.
So what is the best way to fix this code?

(also this is posted in 'Ideas' but its not really an idea; there seems to be no relevant category for code discussion...)

**Comment by bdach** [comment]
@ppy/team-client would be interested in opinions as to whether we can agree with the premise (of exporting without save), and with either of the suggested implementations

**Comment by peppy** [comment]
Taking one step back for a moment: @OliBomby if we are making a lossy export process (where new curve types are lost in the process) I would want a lossless version to also be stored in the exported beatmap, so that on importing to lazer the correct curve types are retained. If we were doing this, you could likely avoid the whole discussion here in the first place, by for instance:

- Always using the legacy conversion pathway for the `.osu` files
- Writing a second `.osu2` or some other extension which stores the lazer version
- Changing lazer to read from the `.osu2` file when parsing beatmaps

The only weird part of doing this is that we will have to handle the scenario where a user exports from lazer, makes changes in stable, then imports back to lazer. This could be done by either:

- Updating stable to delete the `.osu2` file when saving in the editor (or renaming it at least, so it isn't used on re-import)
- Changing lazer to only read from the `.osu2` file if it was modified at the same timestamp as the adjacent `.osu` file

I prefer the first of these two approaches.

---

Keep in mind that as we move forward, the lazer beatmap format is going to diverge further and further from the original one, so having it stored in a separate file is likely beneficial from that angle.

**Reply by OliBomby** [reply]
My idea was to have two explicit export options: one for stable and one for lazer.

- The stable export would apply conversion and make a package compatible with stable.
- The lazer export would not use the conversion and keep a lossless encoding of the beatmap, compatible only with lazer.

The user will choose which export option to use.

I think this is better because it makes explicit to the user that a lossy encoding is happening, which might prevent confusing situations when they notice stuff is different in stable.
Also this is future proof as the file format for lazer and stable further diverge. If for example other files like storyboards change in lazer then we dont need to also add .osb2 and logic to handle those.

**Reply by peppy** [reply]
I guess that's valid too. Not sure which path will work better. Maybe the exports should have something in their filename (the `.osz` filename, that is) to specify whether they are lossless or not, at very least?

**Reply by OliBomby** [reply]
I agree, a different extension name for the new lazer package would make sense.

**Comment by peppy** [comment, marked as answer]
To answer the general question here:

`ExportAsync` already supports `Live<>`. This means that you can already `Detach` an existing beatmap and then make modifications, sending it for export as a `RealmLiveUnmanaged`. The problem with this approach is that you are still confined to providing files from the `UserFileStorage`, even if you can modify the realm objects:

https://github.com/ppy/osu/blob/3f63fd7f4d1f8ee6d87258bb309879e6f540cf8f/osu.Game/Database/LegacyArchiveExporter.cs#L42

If you want to go for the OOP approach, then a path to success may be to rejig `GetStoragePath` to take in a `Storage` and then customise a file model to let you override this. But I think you'll find this to be too complicated, mostly due to realm inheritance limitations.

Easier method would probably be to write the files to the `RealmFileStore`. If you don't write any references to the files, they will be automatically cleaned up on next startup.

so something like..

```csharp
var copy = beatmapInfo.Clone(); // does the detach from realm

copy.BeatmapSet.Files.Add(realmFileStore.Add(stream, realm));

exporter.ExportAsync(new RealmLiveUnmanaged(copy));
```

in theory should be enough.
